File: sdk/samples/all in on code/Visual Studio 2008/VBWinFormBindingNestedProperties/SubPropertyDescriptor.vb
'********************************* Module Header *********************************'
' Module Name:  SubPropertyDescriptor.vb
' Project:      VBWinFormBindToNestedProperties
' Copyright (c) Microsoft Corporation.
' 
' The sample demonstrates how to bind a DataGridView column to a nested property 
' in the data source.
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'*********************************************************************************'

Imports System.ComponentModel

Class SubPropertyDescriptor
    Inherits PropertyDescriptor

    Private _subPD As PropertyDescriptor
    Private _parentPD As PropertyDescriptor

    Public Sub New(ByVal parentPD As PropertyDescriptor, ByVal subPD As PropertyDescriptor, ByVal pdname As String)
        MyBase.New(pdname, Nothing)
        _subPD = subPD
        _parentPD = parentPD
    End Sub

    Public Overrides ReadOnly Property IsReadOnly() As Boolean
        Get
            Return False
        End Get
    End Property
    Public Overrides Sub ResetValue(ByVal component As Object)

    End Sub
    Public Overrides Function CanResetValue(ByVal component As Object) As Boolean
        Return False
    End Function
    Public Overrides Function ShouldSerializeValue(ByVal component As Object) As Boolean
        Return True
    End Function

    Public Overrides ReadOnly Property ComponentType() As Type
        Get
            Return _parentPD.ComponentType
        End Get
    End Property

    Public Overrides ReadOnly Property PropertyType() As Type
        Get
            Return _subPD.PropertyType
        End Get
    End Property

    Public Overrides Function GetValue(ByVal component As Object) As Object
        Return _subPD.GetValue(_parentPD.GetValue(component))
    End Function

    Public Overrides Sub SetValue(ByVal component As Object, ByVal value As Object)
        _subPD.SetValue(_parentPD.GetValue(component), value)
        OnValueChanged(component, EventArgs.Empty)
    End Sub

End Class

File: sdk/samples/all in on code/Visual Studio 2008/CSWinFormBindingNestedProperties/SubPropertyDescriptor.cs
/************************************* Module Header **************************************\
* Module Name:  SubPropertyDescriptor.cs
* Project:      CSWinFormBindToNestedProperties
* Copyright (c) Microsoft Corporation.
* 
* The sample demonstrates how to bind a DataGridView column to a nested property 
* in the data source.
* 
* This source is subject to the Microsoft Public License.
* See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
* All other rights reserved.
* 
* THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
* EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
* WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\******************************************************************************************/

using System;
using System.Collections.Generic;
using System.Text;
using System.ComponentModel;

namespace BindingNestedProperties
{
    class SubPropertyDescriptor : PropertyDescriptor
    {
        private PropertyDescriptor _subPD;
        private PropertyDescriptor _parentPD;

        public SubPropertyDescriptor(PropertyDescriptor parentPD, PropertyDescriptor subPD, string pdname)
            : base(pdname, null)
        {
            _subPD = subPD;
            _parentPD = parentPD;
        }
        
        public override bool IsReadOnly { get { return false; } }

        public override void ResetValue(object component) { }

        public override bool CanResetValue(object component) { return false; }

        public override bool ShouldSerializeValue(object component)
        {
            return true;
        }

        public override Type ComponentType
        {
            get { return _parentPD.ComponentType; }          
        }

        public override Type PropertyType 
        { 
            get { return _subPD.PropertyType; }
        }

        public override object GetValue(object component)
        {
            return _subPD.GetValue(_parentPD.GetValue(component));
        }    

        public override void SetValue(object component, object value)
        {                      
            _subPD.SetValue(_parentPD.GetValue(component), value);
            OnValueChanged(component, EventArgs.Empty);
        }
    }
}